### PR TITLE
Bugfix/1931 requestbody schemas referencing schemas

### LINKED ISSUE #1931: No validation error if request body schema references a requestBody component instead of a schema
### Q&A (please complete the following information)
 - OS: Windows 7
 - Browser: Chrome latest
 - Method of installation: http://editor.swagger.io
 - Swagger-Editor version: 3.6.20
 - Swagger/OpenAPI version: OpenAPI 3.0

### Content & configuration
Example Swagger/OpenAPI definition:
```yaml
openapi: 3.0.0
info:
  title: Invalid request body reference
  version: 0.0.0
paths:
  /foo:
    post:
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/requestBodies/Foo'
      responses:
        '200':
          description: OK

components:
  requestBodies:
    Foo:
      content:
        application/json:
          schema:
            $ref: '#/components/schemas/Foo'
  schemas:
    Foo:
      type: array
      items:
        type: string
```

### Describe the bug you're encountering
In the definition above, `requestBody.content.<media-type>.schema` references a definition from `#/components/requestBodies/` instead of a schema from `#/components/schemas/`. Syntax validation does not flag this.

### Expected behavior
An error about incorrect reference.

### Actual behavior
No error.

### Additional context or thoughts
The root cause might be related to #1717.

### PR DESCRIPTION
**Pull request by ralphdoe**
### Description

This ticket resolves #1931 

### Motivation and Context

This ticket resolves #1931 

### How Has This Been Tested?

Added new unit tests and functional tests have been done.

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [x] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
 export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = () => sys => {
```
Comment (shockey) [inline src/plugins/validate-semantic/validators/oas3/refs.js]: let's drop this leading newline!

```suggestion
export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = () => sys => {
```

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 2]
```diff
@@ -29,3 +30,36 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
       }, [])
     })
 }
+
+  
+export const validateOAS3RefsForRequestBodieSchemasReferenceSchemaPositions = () => sys => {
```
Comment (shockey) [inline src/plugins/validate-semantic/validators/oas3/refs.js]:
```suggestion
export const validateOAS3RequestBodyRefsReferenceAllowableSchemaPositions = () => sys => {
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 3]
```diff
@@ -206,4 +206,84 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+              type: "string"
+            }
+          }
+        }
+      }
+
+      return validateHelper(spec)
+        .then(system => {
+          const allErrors = system.errSelectors.allErrors().toJS()
+          expect(allErrors.length).toEqual(3)
```
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]: **Change request**: let's ignore the first two errors (which aren't semantic errors, and are unrelated to the code under test) by filtering them out.

You could do something like:

```js
const allErrors = system.errSelectors.allErrors().toJS()
const allSematicErrors = allErrors.filter(err => err.source === "semantic")

expect(allSemanticErrors.length).toEqual(1)

const firstError = allSemanticErrors[0]

expect(firstError.message).toEqual(`requestBody schemas $refs must point to a position where a schema can be legally placed`)
expect(firstError.path).toEqual(["paths", "/foo", "post", "requestBody", "content", "application/json", "schema", "$ref"])
```

Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: I tried with this filter, but still having 3 same errors. It seems all 3 are semantic errors.

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 2 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Done!

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 1 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Is this the same as the below comment?

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 4]
```diff
@@ -206,4 +206,82 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+      const spec = {
+        openapi: "3.0.0",
+        paths: {
+          "/foo": {
+            post: {
+              requestBody: {
+                content: {
+                  "application/json": {
+                    schema: {
+                      $ref: "#/components/requestbodies/Foo"
```
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]: @ralphdoe, I believe the two extra errors you're seeing are due to the fact that `#/components/requestbodies/Foo` doesn't actually exist in your spec object — consider changing this to `#/components/requestBodies/Foo` and L230 below from `schemas` to `requestBodies`!

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 5]
```diff
@@ -29,3 +30,36 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+        }
+
+        const [, refPath = ""] = ref.split("#")
+        const pathArr = refPath.split("/") || []
+        const parentRefKey = pathArr.slice(-2)[0]
+        const targetRefKey = pathArr.slice(-1)[0]
+        if(
+          targetRefKey !== "schemas"
+          && parentRefKey !== "schemas"
+          && parentRefKey !== "schemas"
```
Comment (shockey) [inline src/plugins/validate-semantic/validators/oas3/refs.js]: seems like we're making the same statement twice — accidental line?

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 6]
```diff
@@ -206,4 +206,82 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+        }
+      }
+
+      return validateHelper(spec)
+        .then(system => {
+          const allErrors = system.errSelectors.allErrors().toJS()
+          expect(allErrors.length).toEqual(0)
+        })
+    })
+  })
```
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]: **Change request**: please add a test case that ensures remote $refs aren't caught by this validator — e.g. I should be able to do this:

```yaml
openapi: 3.0.0
paths:
  "/foo":
    post:
      requestBody:
        content:
          application/json:
            schema:
              "$ref": "http://google.com#/Foo"
```

... and not get an error, since the validator shouldn't assume anything about how a remote document is structured.

(cc @hkosova — correct me if I'm off track here, this is my current understanding of how the validator is meant to handle remote refs)

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 7]
```diff
@@ -206,4 +206,82 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+          }
+        }
+      }
+
+      return validateHelper(spec)
+        .then(system => {
+          const allErrors = system.errSelectors.allErrors().toJS()
+          const allSemanticErrors = allErrors.filter(err => err.source === "semantic")
+          
+          expect(allSemanticErrors.length).toEqual(3)
```
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]: **Change request**: iron out the other two errors (per the strategy above) so that we only are seeing one error here!

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 8]
```diff
@@ -29,3 +30,36 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+        if(!ref) {
+          return acc
+        }
+
+        const [, refPath = ""] = ref.split("#")
+        const pathArr = refPath.split("/") || []
+        const parentRefKey = pathArr.slice(-2)[0]
+        const targetRefKey = pathArr.slice(-1)[0]
+        if(
+          targetRefKey !== "schemas"
```
Comment (shockey) [inline src/plugins/validate-semantic/validators/oas3/refs.js]: I believe this should be `schema`, based on my read of the validator this was adapted from?

```suggestion
          targetRefKey !== "schema"
```

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 6 above)
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]: **Change request**: please add a test case that ensures local operation $refs aren't caught by this validator, e.g. this should have no problems:

```yaml
openapi: 3.0.0
info:
  title: Enum value validation against schema
  version: 1.0.0
paths: 
  /foo:
    post:
      responses:
        200:
          description: ok
      requestBody:
        content:
          application/json:
            schema:
              type: string
    put:
      responses:
        200:
          description: ok
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/paths/~1foo/post/requestBody/content/application~1json/schema"
```

Comment (hkosova) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: @shockey yes, remote $refs should not be flagged.

Did you mean
```yaml
openapi: 3.0.0
paths:
  "/foo":
    post:
      requestBody:
        "$ref": "http://google.com#/Foo"
```
?

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 5 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Yes, accidental.

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 4 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 7 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Done!

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 8 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 6 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: I implemented a solution for this.

Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Test added!

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 9]
```diff
@@ -29,3 +30,38 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+        const [, refPath = ""] = ref.split("#")
+        const pathArr = refPath.split("/") || []
+        const parentRefKey = pathArr.slice(-2)[0]
+        const targetRefKey = pathArr.slice(-1)[0]
+        if(
+          targetRefKey !== "schema"
+          && parentRefKey !== "schemas"
+          && !ref.startsWith("http://")
+          && !ref.startsWith("https://")
+          && !ref.endsWith(".yaml")
```
Comment (hkosova) [inline src/plugins/validate-semantic/validators/oas3/refs.js]: @ralphdoe a more reliable way to check for local vs external $refs is
```suggestion
          && ref.startsWith("#/")
```
This will exclude all external $refs such as
```
http://...
https://...
path/to/...
./path/to/...
../path/to/...
etc.
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 10]
```diff
@@ -206,4 +206,184 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+                      $ref: "#/components/requestbodies/Foo"
+                    }
+                  }
+                }
+              }
+            }
+          }
+        },
+        components: {
+          requestbodies: {
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js]:
```suggestion
          requestBodies: {
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 11]
```diff
@@ -206,4 +206,184 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+      const spec = {
+        openapi: "3.0.0",
+        paths: {
+          "/foo": {
+            post: {
+              requestBody: {
+                content: {
+                  "application/json": {
+                    schema: {
+                      $ref: "#/components/requestbodies/Foo"
```
Comment (hkosova) [inline test/unit/plugins/validate-semantic/oas3/refs.js]:
```suggestion
                      $ref: "#/components/requestBodies/Foo"
```

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 4 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Done!

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 9 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Thanks Helen!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 10 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Thanks!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 11 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Thanks!

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 1 above)
Comment (shockey) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: it's not! there's a new blank line at the top of this file, and I'd like to remove that 🙂

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 12]
```diff
@@ -206,4 +206,184 @@ describe("validation plugin - semantic - oas3 refs", () => {
       return expectNoErrorsOrWarnings(spec)
     })
   })
+
+  describe("$refs for requestbody schemas must reference a schema by position", () => {
+    it("should return an error when a requestBody schema incorrectly references a local component schema", () => {
```
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]: **change request:** this test case's name is misleading!
```suggestion
    it("should return an error when a requestBody schema incorrectly references a local component requestBody", () => {
```

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 13]
```diff
@@ -29,3 +30,36 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+        const parentRefKey = pathArr.slice(-2)[0]
+        const targetRefKey = pathArr.slice(-1)[0]
+        if(
+          targetRefKey !== "schema"
+          && parentRefKey !== "schemas"
+          && ref.startsWith("#/")
+        ) {
+          acc.push({
+            level: "error",
+            message: `requestBody schemas $refs must point to a position where a schema can be legally placed`,
```
Comment (shockey) [inline src/plugins/validate-semantic/validators/oas3/refs.js]: **change request:** let's sharpen up the messaging here, and update the test cases to match! 🙂 

```suggestion
            message: `requestBody $ref values must point to a position where a Schema Object can be legally placed`,
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 14]
```diff
@@ -206,4 +206,184 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+      }
+
+      return validateHelper(spec)
+        .then(system => {
+          const allErrors = system.errSelectors.allErrors().toJS()
+          expect(allErrors.length).toEqual(0)
+        })
+    })
+
+    it("should not return an error when a requestBody schema references remote references", () => {
```
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]: it's not a good practice to have multiple test cases with the same description — makes it hard to track down what's failing when you see it in reports!

```suggestion
    it("should not return an error when a requestBody schema references entire remote documents", () => {
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 15]
```diff
@@ -206,4 +206,184 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
```
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]:
```suggestion
    it("should not return an error when a requestBody schema references remote document paths", () => {
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 16]
```diff
@@ -206,4 +206,184 @@ describe("validation plugin - semantic - oas3 refs", () => {
...
+      }
+
+      return validateHelper(spec)
+        .then(system => {
+          const allErrors = system.errSelectors.allErrors().toJS()
+          expect(allErrors.length).toEqual(0)
+        })
+    })
+
+    it("should not return an error when a requestBody schema references to local operations", () => {
```
Comment (shockey) [inline test/unit/plugins/validate-semantic/oas3/refs.js]:
```suggestion
    it("should not return an error when a requestBody schema references local operation requestBody schemas", () => {
```

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 13 above)
Comment (hkosova) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: @shockey the 2nd replacement ("schema" -> "Schema Object") is ok, but I disagree about the 1st replacement.

"requestBody schema $refs" != "requestBody $refs". To me, the former refers to `requestBody.content.<media-type>.schema.$ref` whereas the latter refers to `requestBody.$ref`.

I would leave the 1st part of the message as is, except change
requestBody schema**s** $refs
->
requestBody schema $refs
(i.e. singular "schema")

Thoughts/objections?

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 1 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Done! I didn't see it. My bad.

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 12 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 14 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 15 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Done!

File: test/unit/plugins/validate-semantic/oas3/refs.js (hunk 16 above)
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Done!

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 13 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Done! Used: requestBody schema $refs and Schema Object.
